File: gremlin-core/src/main/java/org/apache/tinkerpop/gremlin/language/translator/GroovyTranslateVisitor.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
package org.apache.tinkerpop.gremlin.language.translator;

import org.antlr.v4.runtime.ParserRuleContext;
import org.apache.tinkerpop.gremlin.language.grammar.GremlinParser;
import org.apache.tinkerpop.gremlin.language.grammar.GremlinParser.GenericLiteralVarargsContext;
import org.apache.tinkerpop.gremlin.structure.util.reference.ReferenceVertex;

/**
 * Converts a Gremlin traversal string into a Groovy source code representation of that traversal with an aim at
 * sacrificing some formatting for the ability to compile correctly. The translations may require use of TinkerPop's
 * sugar syntax and therefore requires use of the {@code GremlinLoader} in the gremlin-groovy module unless you are
 * specifically certain that your translations will not result in the use of that syntax. If in doubt, prefer the
 * {@link JavaTranslateVisitor} instead.
 * <ul>
 *     <li>Normalize numeric suffixes to lower case</li>
 *     <li>If floats are not suffixed they will translate as BigDecimal</li>
 *     <li>Makes anonymous traversals explicit with double underscore</li>
 *     <li>Makes enums explicit with their proper name</li>
 * </ul>
 */
public class GroovyTranslateVisitor extends TranslateVisitor {
    private static final String vertexClassName = ReferenceVertex.class.getSimpleName();
    public GroovyTranslateVisitor() {
        this("g");
    }

    public GroovyTranslateVisitor(final String graphTraversalSourceName) {
        super(graphTraversalSourceName);
    }

    @Override
    public Void visitStructureVertex(final GremlinParser.StructureVertexContext ctx) {
        sb.append("new ");
        sb.append(vertexClassName);
        sb.append("(");
        visit(ctx.getChild(3)); // id
        sb.append(", ");
        visit(ctx.getChild(5)); // label
        sb.append(")");
        return null;
    }

    @Override
    public Void visitIntegerLiteral(final GremlinParser.IntegerLiteralContext ctx) {
        final String integerLiteral = ctx.getText().toLowerCase();

        // check suffix
        final int lastCharIndex = integerLiteral.length() - 1;
        final char lastCharacter = integerLiteral.charAt(lastCharIndex);
        switch (lastCharacter) {
            case 'b':
                // parse B/b as byte
                sb.append("(byte)");
                sb.append(integerLiteral, 0, lastCharIndex);
                break;
            case 's':
                // parse S/s as short
                sb.append("(short)");
                sb.append(integerLiteral, 0, lastCharIndex);
                break;
            case 'n':
                // parse N/n as BigInteger which for groovy is "g" shorthand
                sb.append(integerLiteral, 0, lastCharIndex).append("g");
                break;
            default:
                // covers I/i and L/l as Integer and Long respectively
                // everything else just goes as specified
                sb.append(integerLiteral);
                break;
        }
        return null;
    }

    @Override
    public Void visitFloatLiteral(final GremlinParser.FloatLiteralContext ctx) {
        final String floatLiteral = ctx.getText().toLowerCase();

        // check suffix
        final int lastCharIndex = floatLiteral.length() - 1;
        final char lastCharacter = floatLiteral.charAt(lastCharIndex);
        switch (lastCharacter) {
            case 'f':
            case 'd':
                // parse F/f as Float and D/d suffix as Double
                sb.append(floatLiteral, 0, lastCharIndex).append(lastCharacter);
                break;
            case 'm':
                // parse N/n as BigDecimal which for groovy is default for floating point numerics
                if (!floatLiteral.contains(".")) {
                    // 1g is interpreted as BigInteger, so 'g' suffix can't be used here
                    sb.append("new BigDecimal(").append(floatLiteral, 0, lastCharIndex).append("L)");
                } else {
                    sb.append(floatLiteral, 0, lastCharIndex);
                }
                break;
            default:
                // everything else just goes as specified
                sb.append(floatLiteral);
                break;
        }
        return null;
    }

    @Override
    public Void visitInfLiteral(final GremlinParser.InfLiteralContext ctx) {
        if (ctx.SignedInfLiteral().getText().equals("-Infinity")) {
            sb.append("Double.NEGATIVE_INFINITY");
        } else {
            sb.append("Double.POSITIVE_INFINITY");
        }
        return null;
    }

    @Override
    public Void visitNullLiteral(final GremlinParser.NullLiteralContext ctx) {
        if (ctx.getParent() instanceof GremlinParser.GenericLiteralMapNullableArgumentContext) {
            sb.append("null as Map");
            return null;
        }

        sb.append(ctx.getText());
        return null;
    }

    @Override
    public Void visitGenericLiteralSet(GremlinParser.GenericLiteralSetContext ctx) {
        sb.append("[");
        for (int i = 0; i < ctx.genericLiteral().size(); i++) {
            final GremlinParser.GenericLiteralContext genericLiteralContext = ctx.genericLiteral(i);
            visit(genericLiteralContext);
            if (i < ctx.genericLiteral().size() - 1)
                sb.append(", ");
        }
        sb.append("] as Set");
        return null;
    }

    @Override
    public Void visitStringLiteral(final GremlinParser.StringLiteralContext ctx) {
        String literal = ctx.getText();
        literal = literal.replace("$", "\\$");
        sb.append(literal);
        return null;
    }

    @Override
    public Void visitTraversalStrategy(final GremlinParser.TraversalStrategyContext ctx) {
        if (ctx.getChildCount() == 1) {
            sb.append(ctx.getText());
        }
        else {
            // 'new' token is optional for gremlin-lang strategy construction, but required in gremlin-groovy
            if(!ctx.getChild(0).getText().equals("new")) {
                sb.append("new ");
            }
            visitChildren(ctx);
        }
        return null;
    }

    @Override
    public Void visitTraversalSourceSpawnMethod_inject(final GremlinParser.TraversalSourceSpawnMethod_injectContext ctx) {
        return handleInject(ctx);
    }

    @Override
    public Void visitTraversalMethod_inject(final GremlinParser.TraversalMethod_injectContext ctx) {
        return handleInject(ctx);
    }

    @Override
    public Void visitTraversalMethod_hasLabel_String_String(final GremlinParser.TraversalMethod_hasLabel_String_StringContext ctx) {
        // special handling for resolving ambiguous invocations of the hasLabel() step. coerces to the string form
        if (ctx.getChildCount() > 4 && ctx.getChild(2).getText().equals("null")) {
            final GremlinParser.StringLiteralVarargsContext varArgs = (GremlinParser.StringLiteralVarargsContext) ctx.getChild(4);
            sb.append(ctx.getChild(0).getText());
            sb.append("((String) null, ");

            for (int i = 0; i < varArgs.getChildCount(); i += 2) {
                if (varArgs.getChild(i).getText().equals("null")) {
                    sb.append("(String) null");
                } else {
                    visit(varArgs.getChild(i));
                }

                if (i < varArgs.getChildCount() - 1) {
                    sb.append(", ");
                }
            }

            sb.append(")");
            return null;
        }

        return visitChildren(ctx);
    }

    /**
    * Special handling for inject with second `null` argument like g.inject(1, null)
    * inject() ends up being ambiguous with groovy's jdk extension of inject(Object initialValue, Closure closure)
    */
    private Void handleInject(final ParserRuleContext ctx) {
        if (ctx.getChildCount() > 3 && ctx.getChild(2) instanceof GenericLiteralVarargsContext) {
            final GenericLiteralVarargsContext varArgs = (GenericLiteralVarargsContext) ctx.getChild(2);
            if (varArgs.getChildCount() > 2 && "null".equals(varArgs.getChild(2).getText())) {
                sb.append(ctx.getChild(0).getText());
                sb.append("(");
                for (int i = 0; i < varArgs.getChildCount(); i += 2) {
                    if (i == 2) {
                        sb.append("(Object) null");
                    } else {
                        visit(varArgs.getChild(i));
                    }

                    if (i < varArgs.getChildCount() - 1) {
                        sb.append(", ");
                    }
                }

                sb.append(")");
                return null;
            }
        }

        return visitChildren(ctx);
    }
}

File: gremlin-javascript/src/main/javascript/gremlin-javascript/lib/language/translator/GroovyTranslateVisitor.ts
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

// Context parameters are typed as `any` because the generated grammar files
// require accepting arbitrary context types at runtime.

import TranslateVisitor from './TranslateVisitor.js';

/**
 * Converts a Gremlin traversal string into a Groovy source code representation.
 * Mirrors the Java GroovyTranslateVisitor.
 */
export default class GroovyTranslateVisitor extends TranslateVisitor {

    constructor(graphTraversalSourceName: string = 'g') {
        super(graphTraversalSourceName);
    }

    visitIntegerLiteral(ctx: any): void {
        const integerLiteral: string = ctx.getText().toLowerCase();
        const lastChar = integerLiteral[integerLiteral.length - 1];

        if (!/[a-z]/.test(lastChar)) {
            this.sb.push(integerLiteral);
            return;
        }

        const num = integerLiteral.slice(0, -1);
        switch (lastChar) {
            case 'b':
                this.sb.push('(byte)');
                this.sb.push(num);
                break;
            case 's':
                this.sb.push('(short)');
                this.sb.push(num);
                break;
            case 'n':
                // Groovy BigInteger uses the 'g' suffix
                this.sb.push(num);
                this.sb.push('g');
                break;
            default:
                // i and l suffixes and anything else kept as-is
                this.sb.push(integerLiteral);
        }
    }

    visitFloatLiteral(ctx: any): void {
        if (ctx.infLiteral() != null) { this.visit(ctx.infLiteral()); return; }
        if (ctx.nanLiteral() != null) { this.visit(ctx.nanLiteral()); return; }

        const floatLiteral: string = ctx.getText().toLowerCase();
        const lastChar = floatLiteral[floatLiteral.length - 1];

        if (!/[a-z]/.test(lastChar)) {
            this.sb.push(floatLiteral);
            return;
        }

        const num = floatLiteral.slice(0, -1);
        switch (lastChar) {
            case 'f':
            case 'd':
                // keep f/d suffix as-is
                this.sb.push(floatLiteral);
                break;
            case 'm':
                // Groovy BigDecimal: if no decimal point use new BigDecimal(NL), else just the number
                if (!floatLiteral.includes('.')) {
                    this.sb.push('new BigDecimal(');
                    this.sb.push(num);
                    this.sb.push('L)');
                } else {
                    this.sb.push(num);
                }
                break;
            default:
                this.sb.push(floatLiteral);
        }
    }

    visitInfLiteral(ctx: any): void {
        if (ctx.SignedInfLiteral() != null && ctx.SignedInfLiteral().getText() === '-Infinity') {
            this.sb.push('Double.NEGATIVE_INFINITY');
        } else {
            this.sb.push('Double.POSITIVE_INFINITY');
        }
    }

    visitUuidLiteral(ctx: any): void {
        if (ctx.stringLiteral() == null) {
            this.sb.push('UUID.randomUUID()');
            return;
        }
        this.sb.push('UUID.fromString(');
        this.sb.push(ctx.stringLiteral().getText());
        this.sb.push(')');
    }

    visitNullLiteral(ctx: any): void {
        if (ctx.parent?.constructor?.name === 'GenericMapNullableArgumentContext') {
            this.sb.push('null as Map');
        } else {
            this.sb.push(ctx.getText());
        }
    }

    visitGenericSetLiteral(ctx: any): void {
        this.sb.push('[');
        const literals: any[] = ctx.genericLiteral();
        for (let i = 0; i < literals.length; i++) {
            this.visit(literals[i]);
            if (i < literals.length - 1) this.sb.push(', ');
        }
        this.sb.push('] as Set');
    }

    visitStringLiteral(ctx: any): void {
        // Preserve original quote style and escape Groovy GString interpolation
        this.sb.push(ctx.getText().replace(/\$/g, '\\$'));
    }

    visitStringNullableLiteral(ctx: any): void {
        if (ctx.getText() === 'null') {
            this.sb.push('null');
        } else {
            // Preserve original quote style and escape Groovy GString interpolation
            this.sb.push(ctx.getText().replace(/\$/g, '\\$'));
        }
    }

    visitMapKey(ctx: any): void {
        if (ctx.LPAREN() != null && ctx.RPAREN() != null) {
            // Groovy map syntax: expression keys must be parenthesized — preserve the parens
            this.sb.push('(');
            this.visit(ctx.getChild(1));
            this.sb.push(')');
        } else {
            this.visit(ctx.getChild(0));
        }
    }

    visitNakedKey(ctx: any): void {
        // Groovy map/config syntax uses unquoted identifier keys
        this.sb.push(ctx.getText());
    }

    visitKeyword(ctx: any): void {
        const keyword: string = ctx.getText();
        const parentName = ctx.parent?.constructor?.name ?? '';
        if (parentName === 'MapKeyContext' || parentName === 'ConfigurationContext') {
            // Groovy uses unquoted keys
            this.sb.push(keyword);
        } else {
            this.sb.push(keyword);
            this.sb.push(' ');
        }
    }

    visitTraversalStrategy(ctx: any): void {
        if (ctx.getChildCount() === 1) {
            // No-arg strategy: output name as-is (no .instance())
            this.sb.push(ctx.getText());
        } else {
            // Ensure 'new' keyword is present
            if (ctx.getChild(0).getText() !== 'new') {
                this.sb.push('new ');
            }
            this.visitChildren(ctx);
        }
    }

    visitTraversalSourceSpawnMethod_inject(ctx: any): void {
        this.handleInject(ctx);
    }

    visitTraversalMethod_inject(ctx: any): void {
        this.handleInject(ctx);
    }

    visitTraversalMethod_hasLabel_String_String(ctx: any): void {
        if (ctx.getChildCount() > 4 && ctx.getChild(2).getText() === 'null') {
            const varArgs = ctx.getChild(4);
            this.sb.push(ctx.getChild(0).getText());
            this.sb.push('((String) null, ');
            for (let i = 0; i < varArgs.getChildCount(); i += 2) {
                if (varArgs.getChild(i).getText() === 'null') {
                    this.sb.push('(String) null');
                } else {
                    this.visit(varArgs.getChild(i));
                }
                if (i < varArgs.getChildCount() - 1) {
                    this.sb.push(', ');
                }
            }
            this.sb.push(')');
            return;
        }
        this.visitChildren(ctx);
    }

    visitTraversalMethod_option_Merge_Map(ctx: any): void {
        this.sb.push('option(');
        this.visit(ctx.traversalMerge());
        this.sb.push(', ');
        this.tryCastMapNullableArgument(ctx.genericMapNullableArgument());
        this.sb.push(')');
    }

    visitTraversalMethod_mergeV_Map(ctx: any): void {
        this.sb.push('mergeV(');
        this.tryCastMapNullableArgument(ctx.genericMapNullableArgument());
        this.sb.push(')');
    }

    visitTraversalSourceSpawnMethod_mergeV_Map(ctx: any): void {
        this.sb.push('mergeV(');
        this.tryCastMapNullableArgument(ctx.genericMapNullableArgument());
        this.sb.push(')');
    }

    visitTraversalMethod_mergeE_Map(ctx: any): void {
        this.sb.push('mergeE(');
        this.tryCastMapNullableArgument(ctx.genericMapNullableArgument());
        this.sb.push(')');
    }

    visitTraversalSourceSpawnMethod_mergeE_Map(ctx: any): void {
        this.sb.push('mergeE(');
        this.tryCastMapNullableArgument(ctx.genericMapNullableArgument());
        this.sb.push(')');
    }

    private tryCastMapNullableArgument(ctx: any): void {
        const nullableLit = ctx?.genericMapNullableLiteral?.();
        if (nullableLit != null && nullableLit?.nullLiteral?.() != null) {
            this.sb.push('(Map) null');
        } else {
            this.visit(ctx);
        }
    }

    /**
     * Special handling for inject with second `null` argument like g.inject(1, null).
     * inject() is ambiguous in Groovy with jdk extension inject(Object, Closure).
     */
    private handleInject(ctx: any): void {
        if (ctx.getChildCount() > 3) {
            const child2 = ctx.getChild(2);
            if (child2?.constructor?.name === 'GenericLiteralVarargsContext') {
                const varArgs = child2;
                if (varArgs.getChildCount() === 1) {
                    const child0 = varArgs.getChild(0);
                    if (child0?.constructor?.name === 'GenericLiteralExprContext') {
                        const injectArgs = child0;
                        if (injectArgs.getChildCount() > 2 && injectArgs.getChild(2).getText() === 'null') {
                            this.sb.push(ctx.getChild(0).getText());
                            this.sb.push('(');
                            for (let i = 0; i < injectArgs.getChildCount(); i += 2) {
                                if (i === 2) {
                                    this.sb.push('(Object) null');
                                } else {
                                    this.visit(injectArgs.getChild(i));
                                }
                                if (i < injectArgs.getChildCount() - 1) {
                                    this.sb.push(', ');
                                }
                            }
                            this.sb.push(')');
                            return;
                        }
                    }
                }
            }
        }
        this.visitChildren(ctx);
    }
}
